src/main/java/jenkins/plugins/openstack/compute/JCloudsSlaveTemplate.java

```diff
@@ -227,10 +230,59 @@ public boolean canProvision(final Label label) {
         // Ensure predictable server name so we can inject it into user data
         builder.name(serverName);
 
-        if (!Strings.isNullOrEmpty(opts.getImageId())) {
-            String imageId = cloud.getOpenstack().getImageIdFor(opts.getImageId());
-            LOGGER.fine("Setting image id to " + imageId);
-            builder.image(imageId);
+        final Openstack openstack = cloud.getOpenstack();
+        final JCloudsCloud.BootSource bootSource = opts.getBootSource() == null ? JCloudsCloud.BootSource.IMAGE : opts.getBootSource();
+        final String imageNameOrId = opts.getImageId();
+        final String effectiveImageId;
+        if (!Strings.isNullOrEmpty(imageNameOrId)) {
+            final List<String> ids;
+            switch (bootSource) {
+                case IMAGE :
+                    ids = openstack.getImageIdsFor(imageNameOrId);
+                    break;
+                case VOLUMESNAPSHOT :
+                    ids = openstack.getVolumeSnapshotIdsFor(imageNameOrId);
+                    break;
+                default :
+                    LOGGER.config("Don't know how to handle bootSource of " + bootSource + ".");
+                    ids = Collections.EMPTY_LIST;
+                    break;
+            }
+            final int numberFound = ids.size();
+            switch (numberFound) {
+                case 0 :
+                    LOGGER.warning("No active " + bootSource + " '" + imageNameOrId + "' could be found.");
+                    effectiveImageId = null;
+                    break;
+                default :
+                    effectiveImageId = ids.get(numberFound - 1);
+                    LOGGER.warning("Multiple " + bootSource + "s (" + numberFound + ") found with name '" + imageNameOrId
+                            + "'.  Using most recent one, '" + effectiveImageId + "'.");
```

Comment:
It simply came down to a division of responsibility - a caller might want to get all IDs and then decide what to do with them.
In our case here, we want to get just one ID and log a warning about the ambiguity, but you are correct that we could change the semantics so that we always return one ID but log a warning when there's more than one.
Personally, I would prefer issues to do with the configuration of a particular template to come from the JCloudsSlaveTemplate's logger rather than Openstack's logger - in my opinion, the Openstack class should do all the communication that it needs through its API and not require any additional logging - however, as you're the code maintainer here I'll follow your lead on this ;-)

So, just to confirm, you want the "getImageFor" method (and the VolumeSnapshot equivalent) to return "first or null" and log a warning where "first" is not the "only" choice?